core/src/main/java/org/apache/calcite/util/ReflectiveVisitDispatcherImpl.java

```diff
@@ -0,0 +1,134 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to you under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.calcite.util;
+
+import org.apache.calcite.config.CalciteSystemProperty;
+
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
+import com.google.common.collect.ImmutableList;
+
+import java.lang.reflect.Method;
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * Looking up methods relating to reflective visitation. Caching reflect
+ * method globally, every instance of the class shares the global cache to
+ * mitigate java reflection invocation overhead.
+ *
+ * @param <E> Argument type
+ * @param <R> Return type
+ */
+public class ReflectiveVisitDispatcherImpl<R extends ReflectiveVisitor, E>
+    implements ReflectiveVisitDispatcher<R, E> {
+
+  /**
+   * Global singleton cache with limited size used to cache methods. Every instance
+   * of the class will share this cache to mitigate java reflection invocation
+   * overhead when looking up methods.
+   * <p>Note that there should be multiple ways to implement caching, for example,
+   * <p>1) Passing <em>Cache</em> between objects and hopefully using Java's
+   * <em>RAII (Resource Acquisition Is Initialization)</em> like try-finally to
+   * garbage collect.
+   * <p>2) Using <code>ThreadLocal</code> to store cache per thread.
+   * <p>3) Using global cache.
+   * <p>Solution 1) and 2) could introduce complexity to the main course and it only
+   * works per instance or per thread.
+   * Solution 3) which introduces a controllable global small sized cache is a
+   * better solution to balance memory usage and performance.
+   */
+  private static final Cache<List<Object>, Method> GLOBAL_METHOD_CACHE =
```

Comment:
Can we use `LoadingCache`?

Reply:
Yes, good catch, method referencing of CacheLoader.from is more readable and less code is needed.